pkg/loop/plugin_median.go

```diff
@@ -37,116 +39,111 @@ func PluginMedianHandshakeConfig() plugin.HandshakeConfig {
 	}
 }
 
-func PluginMedianClientConfig(lggr logger.Logger) *plugin.ClientConfig {
-	return &plugin.ClientConfig{
-		HandshakeConfig: PluginMedianHandshakeConfig(),
-		Plugins: map[string]plugin.Plugin{
-			PluginMedianName: NewGRPCPluginMedian(nil, lggr),
-		},
-		AllowedProtocols: []plugin.Protocol{plugin.ProtocolGRPC},
-	}
-}
-
-type grpcPluginMedian struct {
+type GRPCPluginMedian struct {
 	plugin.NetRPCUnsupportedPlugin
 
-	impl PluginMedian
-	lggr logger.Logger
-}
+	StopCh <-chan struct{}
+	Logger logger.Logger
 
-func NewGRPCPluginMedian(mp PluginMedian, lggr logger.Logger) plugin.Plugin {
-	return &grpcPluginMedian{impl: mp, lggr: lggr}
+	PluginServer PluginMedian
+
+	pluginClient *pluginMedianClient
 }
 
-func (p *grpcPluginMedian) GRPCServer(broker *plugin.GRPCBroker, server *grpc.Server) error {
-	pb.RegisterPluginMedianServer(server, newPluginMedianServer(&lggrBroker{p.lggr, broker}, p.impl))
-	return nil
+func (p *GRPCPluginMedian) GRPCServer(broker *plugin.GRPCBroker, server *grpc.Server) error {
+	pb.RegisterServiceServer(server, &serviceServer{srv: p.PluginServer})
+	pb.RegisterPluginMedianServer(server, newPluginMedianServer(&lggrBroker{p.StopCh, p.Logger, broker}, p.PluginServer))
+	ctx, cancel := utils.ContextFromChan(p.StopCh)
+	defer cancel()
+	return p.PluginServer.Start(ctx)
 }
 
 // GRPCClient implements [plugin.GRPCPlugin] and returns a [PluginRelayer].
-func (p *grpcPluginMedian) GRPCClient(ctx context.Context, broker *plugin.GRPCBroker, conn *grpc.ClientConn) (interface{}, error) {
-	return newMedianPluginClient(p.lggr, broker, conn), nil
+func (p *GRPCPluginMedian) GRPCClient(_ context.Context, broker *plugin.GRPCBroker, conn *grpc.ClientConn) (interface{}, error) {
+	if p.pluginClient == nil {
+		p.pluginClient = newPluginMedianClient(p.StopCh, p.Logger, broker, conn)
+	} else {
+		p.pluginClient.refresh(broker, conn)
+	}
+
+	return p.pluginClient, nil
+}
+
+func (p *GRPCPluginMedian) ClientConfig() *plugin.ClientConfig {
+	return &plugin.ClientConfig{
+		HandshakeConfig:  PluginMedianHandshakeConfig(),
+		Plugins:          map[string]plugin.Plugin{PluginMedianName: p},
+		AllowedProtocols: []plugin.Protocol{plugin.ProtocolGRPC},
+	}
 }
 
 var _ PluginMedian = (*pluginMedianClient)(nil)
 
 type pluginMedianClient struct {
-	*lggrBroker
+	*pluginClient
```

Comment:
can we give this field names? when you write m.newClientConn on 96, i can't tell which of these two struct is being used

Reply:
This is by design to inherit methods. More succinct access is generally regarded as a feature, but besides that we save on boilerplate re-implementations of wrapper methods.

Reply:
When there is ambiguity the language forces you to be explicit.

Reply:
(the indirect `serviceClient.lb` method is not inherited in this case)
```go
m.pluginClient.newClientConn()
m.serviceClient.lb.newClientConn()
```

Reply:
i'm all for reduced boilerplate.

however, the code has to be maintainable and readable. i've spent 3 1-2 hr sessions reading the code and i simply can't follow what's happening.

an alternative to naming the fields is to add more comments and figure out clearer, more descriptive names. see next comment for an example of my struggles

Reply:
Can you be more specific? Giving up language features or polluting type names don't sound like worthwhile tradeoffs. I could update the `newClientConn` call to be explicit about the field name, but then we're undoing language features to dispel incorrect assumptions that a reader might have. Are there more cases?